tools/pipesys/src/cmd/mod.rs

```diff
@@ -0,0 +1,103 @@
+mod link;
+
+use self::link::Link;
+use pipesys::server::Server as Serve;
+
+use anyhow::{ensure, Context, Result};
+use clap::Parser;
+use env_logger::Builder;
+use log::{debug, LevelFilter};
+use nix::fcntl::{fcntl, F_DUPFD};
+
+const DEFAULT_LEVEL_FILTER: LevelFilter = LevelFilter::Info;
+
+// Don't accept file descriptors 0, 1, or 2 since those correspond to the well-known stdin, stdout,
+// and stderr which could confuse the calling process or its children.
+const MIN_FD: i32 = 3;
+
+/// A tool for passing file descriptors into builds.
+#[derive(Debug, Parser)]
+#[clap(about, long_about = None, version)]
+pub(crate) struct Args {
+    /// Set the logging level. One of [off|error|warn|info|debug|trace]. Defaults to warn. You can
+    /// also leave this unset and use the RUST_LOG env variable. See
+    /// https://github.com/rust-cli/env_logger/
+    #[clap(long = "log-level")]
+    pub(crate) log_level: Option<LevelFilter>,
+
+    #[clap(subcommand)]
+    pub(crate) subcommand: Subcommand,
+}
+
+#[derive(Debug, Parser)]
+pub(crate) enum Subcommand {
+    /// Serve file descriptors to clients.
+    Serve(Serve),
+
+    /// Link a directory file descriptor to the target path.
+    Link(Link),
+}
+
+/// Entrypoint for the `pipesys` command line program.
+pub(super) async fn run(args: Args) -> Result<()> {
+    match args.subcommand {
+        Subcommand::Serve(serve_args) => serve_args.serve().await,
+        Subcommand::Link(link_args) => link_args.execute().await,
+    }
+}
+
+/// use `level` if present, or else use `RUST_LOG` if present, or else use a default.
+pub(super) fn init_logger(level: Option<LevelFilter>) {
+    match (std::env::var(env_logger::DEFAULT_FILTER_ENV).ok(), level) {
+        (Some(_), None) => {
+            // RUST_LOG exists and level does not; use the environment variable.
+            Builder::from_default_env().init();
+        }
+        _ => {
+            // use provided log level or default for this crate only.
+            Builder::new()
+                .filter(
+                    Some(env!("CARGO_CRATE_NAME")),
+                    level.unwrap_or(DEFAULT_LEVEL_FILTER),
+                )
+                .init();
+        }
+    }
+}
+
+/// Helper function to retrieve file descriptors via an abstract socket.
+fn fetch_fds(socket: &str, wanted: usize) -> Result<Vec<i32>> {
+    let addr = uds::UnixSocketAddr::from_abstract(socket.as_bytes())
+        .with_context(|| format!("failed to create socket {}", socket))?;
```

Comment:
Seems like `wanted` doesn't actually behave like expected. Should probably alter the function or else just always fetch one FD as a constant.